jfxutils/src/main/java/org/gillius/jfxutils/chart/ChartZoomManager.java

```diff
@@ -502,19 +527,112 @@ public void handle( ScrollEvent event ) {
 				double zoomAmount = 0.2 * direction;
 
 				if ( zoomMode == AxisConstraint.Both || zoomMode == AxisConstraint.Horizontal ) {
-					double xZoomDelta = ( xAxis.getUpperBound() - xAxis.getLowerBound() ) * zoomAmount;
+					double xZoomDelta = ( getXAxisUpperBound() - getXAxisLowerBound() ) * zoomAmount;
 					xAxis.setAutoRanging( false );
-					xAxis.setLowerBound( xAxis.getLowerBound() - xZoomDelta * xZoomBalance );
-					xAxis.setUpperBound( xAxis.getUpperBound() + xZoomDelta * ( 1 - xZoomBalance ) );
+					setXAxisLowerBound( getXAxisLowerBound() - xZoomDelta * xZoomBalance );
+					setXAxisUpperBound( getXAxisUpperBound() + xZoomDelta * ( 1 - xZoomBalance ) );
 				}
 
 				if ( zoomMode == AxisConstraint.Both || zoomMode == AxisConstraint.Vertical ) {
-					double yZoomDelta = ( yAxis.getUpperBound() - yAxis.getLowerBound() ) * zoomAmount;
+					double yZoomDelta = ( getYAxisUpperBound() - getYAxisLowerBound() ) * zoomAmount;
 					yAxis.setAutoRanging( false );
-					yAxis.setLowerBound( yAxis.getLowerBound() - yZoomDelta * yZoomBalance );
-					yAxis.setUpperBound( yAxis.getUpperBound() + yZoomDelta * ( 1 - yZoomBalance ) );
+					setYAxisLowerBound( getYAxisLowerBound() - yZoomDelta * yZoomBalance );
+					setYAxisUpperBound( getYAxisUpperBound() + yZoomDelta * ( 1 - yZoomBalance ) );
 				}
 			}
 		}
 	}
+
+	private double getXAxisLowerBound() {
+		return xAxisLowerBoundProperty.get();
+	}
+
+	private void setXAxisLowerBound(double val) {
+		xAxisLowerBoundProperty.set(val);
+	}
+
+	private double getXAxisUpperBound() {
+		return xAxisUpperBoundProperty.get();
+	}
+
+	private void setXAxisUpperBound(double val) {
+		xAxisUpperBoundProperty.set(val);
+	}
+
+	private double getYAxisLowerBound() {
+		return yAxisLowerBoundProperty.get();
+	}
+
+	private void setYAxisLowerBound(double val) {
+		yAxisLowerBoundProperty.set(val);
+	}
+
+	private double getYAxisUpperBound() {
+		return yAxisUpperBoundProperty.get();
+	}
+
+	private void setYAxisUpperBound(double val) {
+		yAxisUpperBoundProperty.set(val);
+	}
+
+	private <T> DoubleProperty getLowerBoundProperty(Axis<T> axis) {
+		return axis instanceof ValueAxis ?
+				((ValueAxis<?>) axis).lowerBoundProperty() :
+				toDoubleProperty(axis, getProperty(axis, "lowerBoundProperty"));
+	}
+
+	private <T> DoubleProperty getUpperBoundProperty(Axis<T> axis) {
+		return axis instanceof ValueAxis ?
+				((ValueAxis<?>) axis).upperBoundProperty() :
+				toDoubleProperty(axis, getProperty(axis, "upperBoundProperty"));
+	}
+
+	private <T> DoubleProperty toDoubleProperty(Axis<T> axis, Property<T> property) {
+		final StringProperty stringProperty = new SimpleStringProperty();
+		final DoubleProperty doubleProperty = new SimpleDoubleProperty() {
+			// keep a reference so that the stringProperty doesn't get garbage-collected
+			private final StringProperty stringProp = stringProperty;
+		};
+
+		stringProperty.bindBidirectional(
```

Comment:
Why do you need the intermediate string property when toNumericValue and toRealValue already convert to and from double? Why can't the doubleProperty bind directly to the property?